flink-core/src/main/java/org/apache/flink/api/common/time/Deadline.java

```diff
@@ -103,11 +107,23 @@ public static Deadline fromNow(Duration duration) {
      * @param clock Time provider for this deadline.
      */
     public static Deadline fromNowWithClock(Duration duration, Clock clock) {
-        return new Deadline(Math.addExact(clock.relativeTimeNanos(), duration.toNanos()), clock);
+        return new Deadline(
+                addHandlingOverflow(clock.relativeTimeNanos(), duration.toNanos()), clock);
     }
 
     @Override
     public String toString() {
         return LocalDateTime.now().plus(timeLeft()).toString();
     }
+
+    // -------------------- private helper methods ----------------
+
+    private static long addHandlingOverflow(long x, long y) {
```

Comment:
nit: we could add a comment it has been copied over from `Math#addExact` and we changed the overflow behaviour.